src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -34,52 +34,72 @@ def self.compile_ast(code, compile_options, boltlib_path)
+                                                                      compile_options['project_root'],
+                                                                      true)
+        # TODO: Given we hide this from plan authors this current iteration has only
+        # the "required" data for now. Once we can get https://github.com/puppetlabs/bolt/pull/1770
+        # merged and promoted we can just use an empty hash.
+        fake_config = {
+          'transport' => 'redacted',
+          'transports' => {
+            'redacted' => 'redacted'
+          }
+        }
+        bolt_inv = Bolt::ApplyInventory.new(fake_config)
+        puppet_overrides = {
+          bolt_inventory: bolt_inv,
+          bolt_project: bolt_project,
+        }
 
-          target_variables = compile_options.dig('target_variables', 'values') || {}
+        # Prior to PE-29443 variables were in a hash. Serialization between ruby/clojure/json did
+        # not preserver hash order necessary for deserialization. The data strucutre is now stored
+        # in a list for moving data and the list is used to construct an ordered ruby hash.
+        plan_variables = if compile_options['variables']['values'].is_a?(Array)
+          compile_options['variables']['values'].each_with_object({}) do |param_hash, acc|
+            acc[param_hash.keys.first] = param_hash.values.first
+          end
+        else
+          compile_options['variables']['values']
+        end
 
-          variables = {
-            variables: plan_variables,
-            target_variables: target_variables,
-          }
+        target_variables = compile_options.dig('target_variables', 'values') || {}
 
-          env_conf = {
-            pre_modulepath: boltlib_path,
-            envpath: Puppet[:environmentpath],
-            facts: compile_options['facts']['values'],
-          }
+        variables = {
+          variables: plan_variables,
+          target_variables: target_variables,
+        }
 
-          # Use the existing environment with the requested name
-          Puppet::Pal.in_environment(compile_options['environment'], env_conf) do |pal|
-            # TODO: Given we hide this from plan authors this current iteration has only
-            # the "required" data for now. Once we can get https://github.com/puppetlabs/bolt/pull/1770
-            # merged and promoted we can just use an empty hash.
-            fake_config = {
-              'transport' => 'redacted',
-              'transports' => {
-                'redacted' => 'redacted'
-              }
-            }
-            bolt_inv = Bolt::ApplyInventory.new(fake_config)
-            Puppet.override(bolt_inventory: bolt_inv) do
+        env_conf = {
+          pre_modulepath: boltlib_path,
+          modulepath: compile_options['modulepath'],
+          facts: compile_options['facts']['values'],
+        }
+
+        puppet_settings = {
+          node_name_value: target['name'],
```

Comment:
This looks like a copy paste issue from bolt (`target` is undefined in this context). It looks like we set this (for env plans) like: `Puppet[:node_name_value] = compile_options['certname']`. Maybe we should update environments to do the override like we do here for hiera config. I'm not sure what the implications of doing puppet settings are like that. 

https://github.com/puppetlabs/puppetserver/pull/2555/files#diff-6f7f4ebbd198c182ba9b89fe43839b117bc1899b5bb3f7ffeddcd1cc64cd79abR119

Reply:
There shouldn't be any implications for changing things to do it the bolt way.

I just realized we _can't_ use the `Puppet[:node_name_value] = ` directly because we completely wipe the settings between assigning that setting and performing the compile. For project compiles we _have_ to pass the value through the puppet_settings hash